src/Features/Core/Portable/MetadataAsSource/VirtualMetadataDocumentPersister.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MetadataAsSource;
+
+internal class VirtualMetadataDocumentPersister() : IMetadataDocumentPersister
+{
+    public const string VirtualFileScheme = "roslyn-metadata";
+    public void CleanupGeneratedDocuments()
+    {
+        // Documents will be cleaned up by the workspace, nothing else we need to do here.
+        return;
+    }
+
+    public string ConvertFilePathToDocumentPath(Guid identifier, string providerName, string filePath)
+    {
+        var fileName = Path.GetFileName(filePath);
+        return GenerateDocumentPath(identifier, providerName, fileName);
+    }
+
+    public string GenerateDocumentPath(Guid identifier, string providerName, string fileName)
+    {
+        return $"{VirtualFileScheme}://{providerName}/{identifier:N}/{fileName}";
```

Comment:
do we own the providername, ensuring that this URI is ok?

similarly, is it safe for the 'filename' to just be concatted to the uri like ths?

Reply:
> do we own the providername, ensuring that this URI is ok?

yes, its the type name of the decompilation or pdb provider

> similarly, is it safe for the 'filename' to just be concatted to the uri like ths?

It should be.  For metadata we generate it, for PDB we either generate it or it comes from the debugger as a valid file name

> do we need filename+guid? is the guid not enough to uniquely identify this? what purpose doe sthe filename serve? is it more just to help with logging? if so, can you mention that that's the case?

Guid is not always enough to identify it - in the PDB case it is the projectId.  We also need the file name as it is rendered in the editor as the document name.

Reply:
> yes, its the type name of the decompilation or pdb provider

Should "provider name" just take the provider object and then GetType().Name or something to get it? That way it'll be a bit more strongly typed where this provider name came from.